internal/cloud/acr/acr.go

```diff
@@ -0,0 +1,164 @@
+package acr
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/services/preview/containerregistry/runtime/2019-08-15-preview/containerregistry"
+	"github.com/Azure/go-autorest/autorest"
+	"github.com/Azure/go-autorest/autorest/adal"
+	"github.com/Azure/go-autorest/autorest/azure/auth"
+	"github.com/Azure/go-autorest/autorest/to"
+	"github.com/docker/docker/api/types"
+	"github.com/dominodatalab/forge/internal/cloud"
+	"github.com/go-logr/logr"
+)
+
+// https://github.com/Azure/acr/blob/main/docs/AAD-OAuth.md
+
+const acrUserForRefreshToken = "00000000-0000-0000-0000-000000000000"
+
+var (
+	acrRegex = regexp.MustCompile(`.*\.azurecr\.io|.*\.azurecr\.cn|.*\.azurecr\.de|.*\.azurecr\.us`)
+)
+
+type authDirective struct {
+	service string
+	realm   string
+}
+
+type acrProvider struct {
+	logger                logr.Logger
+	tenantID              string
+	servicePrincipalToken *adal.ServicePrincipalToken
+}
+
+func Register(logger logr.Logger, registry *cloud.Registry) error {
+	provider, err := newProvider(logger)
+	if err != nil {
+		return err
+	}
+
+	registry.Register(acrRegex, provider.authenticate)
+	return nil
+}
+
+func newProvider(logger logr.Logger) (*acrProvider, error) {
+	settings, err := auth.GetSettingsFromEnvironment()
+	if err != nil {
+		return nil, err
+	}
+
+	// the minimum set of required values
+	if settings.Values[auth.TenantID] == "" || settings.Values[auth.ClientID] == "" {
+		return nil, errors.New("no Azure credentials")
+	}
+
+	var spt *adal.ServicePrincipalToken
+	if cc, err := settings.GetClientCredentials(); err == nil {
+		if spt, err = cc.ServicePrincipalToken(); err != nil {
+			return nil, err
+		}
+	} else {
+		ctx := context.Background()
+		for i := 0; i < 3; i++ {
+			if spt, err = settings.GetMSI().ServicePrincipalToken(); err == nil {
+				break
+			}
+			logger.Error(err, "retrying", "attempt", i+1)
+			if !autorest.DelayForBackoff(time.Second, i, ctx.Done()) {
+				return nil, ctx.Err()
+			}
+		}
+		if err != nil {
+			// IMDS can take some time to setup, restart the process
+			panic(err)
```

Comment:
Why should the error matter? 
And this begins the issue. 
In the case you provide a full set of credentials, we don't validate them. This will appear as a failure when attempting to process a CIB unless we attempt to obtain a token during registration which might be a good idea for consistency.
When using MSI, we can fail to obtain the credentials because the information is invalid or the MSI endpoint is not setup yet. For either, we should exit given you intend to use MSI and its failing.

Reply:
>  When using MSI, we can fail to obtain the credentials because the information is invalid or the MSI endpoint is not setup yet. For either, we should exit given you intend to use MSI and its failing.

The name of the error doesn't really matter to me as much as the message. I would assume that, as a consumer of this API, if the MSI endpoint is not setup, then tell me so I can figure out what to do about it. It just so happens we want to exist but I think the controller clearly exiting because MSI is not ready is more obvious than a `panic` causing the process to crash exit. 

Does that make sense?

Reply:
Welcome to Azure. The issue isn't whether its setup or not. It's a timing issue with Azure. MSI can be expected but take seconds after the instance is up before it's connected.  The docs for MSI even state that you need to retry if you get a Connection Refused which is how it appears.
I have a slightly different philosophy, if you have configured something and it fails, do not pass go.
If you provide details for MSI and it fails for whatever reason, error out. Let the user decide to either remove the configuration or fix whatever is wrong. Doing anything else means we start without ACR when you want it and it will surface as errors in Domino which aren't obvious.